### PR TITLE
ENH: Pass all arguments as kwargs

### PR DESCRIPTION
**Pull request by peterbell10**
Implements the first part of #164 

All positional arguments are converted into a dictionary and used to update `kwargs`. varargs are stored in the `kwargs` dictionary as a tuple e.g.

```python
def min(a, *args):
	pass
```
gives kwargs of the form
```python
{'a', a, 'args': (arg1, arg2, ...)}
```

### CODE REVIEW

File: docs/libauthor_docs.rst [hunk 1]
```diff
@@ -26,12 +26,13 @@ both use the ``numpy`` domain.
...
-Note that it is called in this form, so if your backend is an object instead of
-a module, you should add ``self``. ``method`` is the multimethod being called,
-and it is guaranteed that it is in the same domain as the backend. ``args`` and
-``kwargs`` are the arguments to the function, possibly after conversion
-(explained below)
+multimethod. It has the signature ``(method, kwargs)``. Note that it is called
+in this form, so if your backend is an object instead of a module, you should
+add ``self``. ``method`` is the multimethod being called, and it is guaranteed
+that it is in the same domain as the backend. ``kwargs`` is a ``dict``
+containging all arguments to the function, possibly after conversion (explained
```
Comment (hameerabbasi) [inline docs/libauthor_docs.rst]:
```suggestion
containing all arguments to the function, possibly after conversion (explained
```

File: unumpy/multimethods.py [hunk 2]
```diff
@@ -506,56 +495,49 @@ def broadcast_arrays(*args, subok=False):
...
 
-def _first_argreplacer(args, kwargs, arrays1):
-    def func(arrays, axis=0, out=None):
-        return (arrays1,), dict(axis=0, out=None)
-
-    return func(*args, **kwargs)
-
-
-@create_numpy(_first_argreplacer)
+@create_numpy(_generic_argreplacer('arrays'))
```
Comment (hameerabbasi) [inline unumpy/multimethods.py]: See comment for `stack`.

File: unumpy/multimethods.py [hunk 3]
```diff
@@ -506,56 +495,49 @@ def broadcast_arrays(*args, subok=False):
...
 @all_of_type(ndarray)
 def concatenate(arrays, axis=0, out=None):
     return arrays
 
 
-@create_numpy(_first_argreplacer)
+@create_numpy(_generic_argreplacer('arrays'))
```
Comment (hameerabbasi) [inline unumpy/multimethods.py]: Here, you're replacing the whole argument with one array. You must replace `arrays` with a list of all passed-in arrays.

File: unumpy/multimethods.py [hunk 4]
```diff
@@ -497,7 +488,8 @@ def lexsort(keys, axis=None):
 
 
 def _args_argreplacer(args, kwargs, arrays):
```
Comment (hameerabbasi) [inline unumpy/multimethods.py]:
```suggestion
def _args_argreplacer(kwargs, arrays):
```

**Comment by peterbell10** [comment]
>Looks like unfortunately NumPy _expects_ posargs, and so without a small fix to `__ua_function__` to move a or self to posargs, this won't work.

I don't follow, can you give an example?

**Comment by hameerabbasi** [comment]
```python
   def __ua_function__(method, kwargs):
        args = ()
        if 'self' in kwargs:
            args = args + (kwargs.pop('self'),)
        if 'a' in kwargs:
            args = args + (kwargs.pop('a'),)
        if method in _implementations:
            return _implementations[method](*args, **kwargs)

        if not hasattr(cp, method.__name__):
            return NotImplemented

        return getattr(cp, method.__name__)(**kwargs)
```

**Comment by peterbell10** [comment]
I think it should be work without that, it's just that the argument names have to match now.

**Comment by hameerabbasi** [comment]
> I think it should be work without that, it's just that the argument names have to match now.

That's a fair point. 🙂

**Comment by hameerabbasi** [comment]
Looks like Dask and Sparse still have issues. 🙍‍♂

**Comment by hameerabbasi** [comment]
I wonder if all this trouble is worth it... It seemed like a good idea in theory. I would like to propose the following change:

* We keep around args, but normalize so that any default args are removed if they're at the end of `*args`.

**Comment by peterbell10** [comment]
Yes, it looks like the different libraries use different names for some parameters. In python 3.8, these would probably be marked as [positional-only parameters][PEP].

How about adding a parameter to `generate_multimethod` that gives the number of positional-only parameters. That way positional-only parameters can still be defaulted.

[PEP]: https://www.python.org/dev/peps/pep-0570/

**Comment by hameerabbasi** [comment]
I'm proposing something different, but a bit more magical: We store the defaults by position as well as by keyword. Then, we trim `*args` from the end.

**Comment by peterbell10** [comment]
I understood, I am just suggesting an alternative :)

If the argument is conceptually positional-only then I think that the current backends will not work properly when they're passed as `kwargs` so it may be worth considering.

**Comment by hameerabbasi** [comment]
> I understood, I am just suggesting an alternative :)

It's a valid alternative. When Python 3.8 comes along and is widely used, we can probably move to it... But in the meanwhile, adding an extra parameter adds overhead for API authors that they may not be willing to foot.

**Comment by hameerabbasi** [comment]
Would you like to keep this open for the future?

**Comment by peterbell10** [comment]
I'll just close for now.

**Pull request closed without merging**